java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -34,73 +38,154 @@ protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
     this.v2 = v2;
   }
 
-  public Boolean impliedBy(List<SymbolicValueRelation> knownRelations) {
+  @Override
+  public int compareTo(SymbolicValueRelation o) {
+    final int[] key1 = key();
+    final int[] key2 = o.key();
+    int answer = Integer.compare(key1[0], key2[0]);
+    if (answer == 0) {
+      answer = Integer.compare(key1[1], key2[1]);
+      if (answer == 0) {
+        answer = o.getClass().getSimpleName().compareTo(o.getClass().getSimpleName());
+      }
+    }
+    return answer;
+  }
+
+  private int[] key() {
+    if (v1.id() < v2.id()) {
+      return new int[] {v1.id(), v2.id()};
+    } else {
+      return new int[] {v2.id(), v1.id()};
+    }
+  }
+
+  @CheckForNull
+  public Boolean impliedBy(Collection<SymbolicValueRelation> knownRelations) {
+    return impliedBy(knownRelations, new TreeSet<SymbolicValueRelation>());
+  }
+
+  @CheckForNull
+  public Boolean impliedBy(Collection<SymbolicValueRelation> knownRelations, Set<SymbolicValueRelation> usedRelations) {
+    if (knownRelations.isEmpty()) {
+      return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Null is returned but a "Boolean" is expected. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2447)